### ISSUE TITLE
Threadview

### ISSUE THREAD

**Issue by romanofski** [purebred-mua/purebred#35]
Apart from changing the order of the list of mails, which should be mostly straight forward, showing threads will be a bit trickier since it needs some additional information for drawing.

I'm wondering if the sorting order (including showing threads) needs a special data type to make this easier "configurable".

**Comment by frasertweedale** [comment]
notmuch has thread-grouping built in, i.e. iterate threads, then iterate mails within thread. 

OTOH it may be more consistent to do the thread grouping ourselves and/or make it completely
generic how to group/sort mail.  But we definitely need thread-based view.

For now I'd say just use the notmuch capability and group by thread without worrying about intra-thread heirarchy of messages.  We will have to construct that ourselves using *Message-Id* data
but we can probably kick that can down the road a bit.

**Comment by romanofski** [comment]
True. #36 is for sorting so I renamed this one, since I think we need a different widget. Maybe good opportunity to contribute it to brick.

**Comment by frasertweedale** [comment]
Yeah its kind of like a "tree list" concept.  Could have folding capability built in (perhaps not as an initial requirement).

The basic representation of a thread would be a rose tree (e.g. `Data.Tree.Tree`) of `Mail`:

```haskell
import Data.Tree (Tree(..))  -- data Tree a = Node a [Tree a]

-- email datatype
data Mail

type Thread = Tree Mail
```

Now, each thread starts from a single email, *but* a user's mail database might not have that email,
esentially leaving you with a `[Tree Mail]` but without the original `Mail`.  So the required representation
would actually be:

```haskell
-- a non-empty list of "sibling" mails
--  (in the common case, exactly one - the original email)
-- each with their own possibly-empty tree of replies
type Thread = NonEmpty (Tree Mail)
```
Each "sibling" mail would be rendered as linked somehow.  e.g. in mutt, when multiple emails are `In-Reply-To` some message that is not available, it is rendered like:
```
     30   C 10/12 Roman Joost     (8.3K) ┬─>Re: [purebred-mua/purebred] Some minor refactoring of event handling (#77) 
     31   C 10/13 Roman Joost     (9.3K) └─>Re: [purebred-mua/purebred] Some minor refactoring of event handling (#77) 
```

So basically we want a widget to display that data :)

**Comment by frasertweedale** [comment]
@romanofski should we close this in light of the current UI paradigm?

**Comment by romanofski** [comment]
Yeh lets close it and open another one if we have new ideas.
